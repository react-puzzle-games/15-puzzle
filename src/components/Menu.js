import React, { Component } from 'react';
import styled from 'styled-components';
import { Toolbar, ToolbarGroup, ToolbarTitle } from 'material-ui/Toolbar';
import RaisedButton from 'material-ui/RaisedButton';
import Chip from 'material-ui/Chip';
import Avatar from 'material-ui/Avatar';
import Alarm from 'material-ui/svg-icons/action/alarm';
import Moves from 'material-ui/svg-icons/action/compare-arrows';
import Replay from 'material-ui/svg-icons/av/replay';
import MediaQuery from 'react-responsive';
import GitHubIcon from './GitHubIcon';

const StyledToolbar = styled(Toolbar)`

@media (max-width: 1094px) {
  
  & {
    justify-content: center !important;
  }

  .toolbarTitle {
    display: none;
  }

}

@media (max-width: 578px) {

  .menuButton {
    margin: 10px 5px !important;
  }

  .menuIcon {
    margin-left: 0px !important;
  }

  .menuIcon+span {
    display: none !important;
  }

}

`;

class Menu extends Component {
  render() {
    const { seconds, moves, onResetClick } = this.props;

    return (
      <StyledToolbar className="toolbar">
        <ToolbarTitle
          className="toolbarTitle"
          text="React Puzzle Games - 15 Puzzle"
        />
        <ToolbarGroup>
          <RaisedButton
            className="menuButton"
            label="Project home"
            target="_blank"
            href="https://github.com/react-puzzle-games/15-puzzle/"
            icon={<GitHubIcon className="menuIcon" />}
          />
          <RaisedButton
            className="menuButton"
            label="Reset game"
            onTouchTap={onResetClick}
            icon={<Replay className="menuIcon" />}
          />
          <Chip>
            <Avatar icon={<Alarm />} />
            <MediaQuery query="(min-width: 772px)" component="span">
              Time Elapsed:{' '}
            </MediaQuery>
            {seconds}s
          </Chip>
          <Chip>
            <Avatar icon={<Moves />} />
            <MediaQuery query="(min-width: 772px)" component="span">
              Moves Counter:{' '}
            </MediaQuery>
            {moves}
          </Chip>
        </ToolbarGroup>
      </StyledToolbar>
    );
  }
}

export default Menu;
