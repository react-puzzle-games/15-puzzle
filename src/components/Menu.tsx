import {
  AppBar,
  Avatar,
  Button,
  Chip,
  Toolbar,
  Typography,
} from "@material-ui/core";
import { makeStyles } from "@material-ui/core/styles";
import {
  Alarm,
  CompareArrows,
  Pause,
  PlayArrow,
  PowerSettingsNew,
  Replay,
} from "@material-ui/icons";
import React from "react";
import MediaQuery from "react-responsive";
import { GameStatus, GAME_PAUSED, GAME_STARTED } from "../lib/game-status";

const useStyles = makeStyles({
  root: {
    backgroundColor: "rgb(232, 232, 232)",
  },

  toolbar: {
    ["@media (max-width: 414px)"]: {
      justifyContent: "center",
    },
  },

  title: {
    color: "#000",
    flexGrow: 1,
  },
});

type Props = {
  seconds: number;
  moves: number;
  onResetClick: () => void;
  onPauseClick: () => void;
  onNewClick: () => void;
  gameState: GameStatus;
};

const Menu = (props: Props) => {
  const {
    seconds = 0,
    moves = 0,
    onResetClick,
    onPauseClick,
    onNewClick,
    gameState,
  } = props;
  const classes = useStyles(props);

  return (
    <AppBar position="static" className={classes.root}>
      <Toolbar className={classes.toolbar}>
        <MediaQuery query="(min-width: 772px)">
          <Typography className={classes.title} variant="h6" component="div">
            React Puzzle Games - 15 Puzzle
          </Typography>
        </MediaQuery>

        <Button
          aria-label="Start a new game"
          onClick={onNewClick}
          startIcon={<PowerSettingsNew className="menuIcon" />}
        >
          <MediaQuery query="(min-width: 772px)">
            <Typography component="span" variant="button">
              New game
            </Typography>
          </MediaQuery>
        </Button>
        <Button
          aria-label="Pause/Continue current game."
          onClick={onPauseClick}
          startIcon={
            gameState === GAME_PAUSED ? (
              <PlayArrow className="menuIcon" />
            ) : (
              <Pause className="menuIcon" />
            )
          }
          disabled={gameState !== GAME_STARTED && gameState !== GAME_PAUSED}
        >
          <MediaQuery query="(min-width: 772px)">
            <Typography component="span" variant="button">
              {gameState === GAME_PAUSED ? "Continue" : "Pause"}
            </Typography>
          </MediaQuery>
        </Button>
        <Button
          aria-label="Reset game"
          onClick={onResetClick}
          startIcon={<Replay />}
        >
          <MediaQuery query="(min-width: 772px)" component="span">
            Reset game
          </MediaQuery>
        </Button>
        <Chip
          avatar={
            <Avatar>
              <Alarm />
            </Avatar>
          }
          label={
            <>
              <MediaQuery query="(min-width: 772px)" component="span">
                Time Elapsed:
              </MediaQuery>
              <Typography component="span">{seconds}s</Typography>
            </>
          }
        />
        <Chip
          avatar={
            <Avatar>
              <CompareArrows />
            </Avatar>
          }
          label={
            <>
              <MediaQuery query="(min-width: 772px)" component="span">
                Moves so far:
              </MediaQuery>
              <Typography component="span">{moves}</Typography>
            </>
          }
        />
      </Toolbar>
    </AppBar>
  );
};

export default Menu;
