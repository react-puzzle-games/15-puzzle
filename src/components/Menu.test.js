import renderer from 'react-test-renderer';
import React from 'react';
import Menu from './Menu';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';

describe('Menu renders correctly', () => {
  test('given default values', () => {
    const tree = renderer
      .create(
        <MuiThemeProvider>
          <Menu seconds={0} moves={0} onResetClick={() => {}} />
        </MuiThemeProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
