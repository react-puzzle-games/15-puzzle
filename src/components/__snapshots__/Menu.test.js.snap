// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders correctly given default values 1`] = `
<div
  className="toolbar sc-bwzfXH chaHyI"
  style={
    Object {
      "WebkitTapHighlightColor": "rgba(0,0,0,0)",
      "backgroundColor": "rgb(232, 232, 232)",
      "boxSizing": "border-box",
      "display": "flex",
      "height": 56,
      "justifyContent": "space-between",
      "muiPrepared": true,
      "padding": "0px 24px",
    }
  }
>
  <span
    className="toolbarTitle"
    style={
      Object {
        "fontFamily": "Roboto, sans-serif",
        "fontSize": 20,
        "lineHeight": "56px",
        "muiPrepared": true,
        "overflow": "hidden",
        "paddingRight": 16,
        "position": "relative",
        "textOverflow": "ellipsis",
        "whiteSpace": "nowrap",
      }
    }
  >
    React Puzzle Games - 15 Puzzle
  </span>
  <div
    className={undefined}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "space-between",
        "marginLeft": undefined,
        "marginRight": undefined,
        "muiPrepared": true,
        "position": "relative",
      }
    }
  >
    <div
      className="menuButton"
      style={
        Object {
          "WebkitTapHighlightColor": "rgba(0,0,0,0)",
          "backgroundColor": "#ffffff",
          "borderRadius": 2,
          "boxShadow": "0 1px 6px rgba(0, 0, 0, 0.12),
                 0 1px 4px rgba(0, 0, 0, 0.12)",
          "boxSizing": "border-box",
          "color": "rgba(0, 0, 0, 0.87)",
          "display": "inline-block",
          "fontFamily": "Roboto, sans-serif",
          "margin": "10px 24px",
          "minWidth": 88,
          "muiPrepared": true,
          "position": "relative",
          "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
        }
      }
    >
      <a
        disabled={false}
        href="https://github.com/react-puzzle-games/15-puzzle/"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        onTouchTap={[Function]}
        style={
          Object {
            "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
            "backgroundColor": "#ffffff",
            "border": 10,
            "borderRadius": 2,
            "boxSizing": "border-box",
            "cursor": "pointer",
            "display": "inline-block",
            "fontFamily": "Roboto, sans-serif",
            "fontSize": "inherit",
            "fontWeight": "inherit",
            "height": 36,
            "lineHeight": "36px",
            "margin": 0,
            "muiPrepared": true,
            "outline": "none",
            "padding": 0,
            "position": "relative",
            "textAlign": "center",
            "textDecoration": "none",
            "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
            "verticalAlign": "middle",
            "width": "100%",
            "zIndex": 1,
          }
        }
        tabIndex={0}
        target="_blank"
      >
        <div
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
        >
          <div
            style={
              Object {
                "backgroundColor": false,
                "borderRadius": 2,
                "height": 36,
                "muiPrepared": true,
                "top": 0,
                "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              }
            }
          >
            <svg
              className="menuIcon sc-bdVaJa dSoADu"
              height="24px"
              version="1.1"
              viewBox="0 0 435.549 435.549"
              width="24px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M409.132,114.573c-19.608-33.596-46.205-60.194-79.798-79.8C295.736,15.166,259.057,5.365,219.271,5.365 c-39.781,0-76.472,9.804-110.063,29.408c-33.596,19.605-60.192,46.204-79.8,79.8C9.803,148.168,0,184.854,0,224.63 c0,47.78,13.94,90.745,41.827,128.906c27.884,38.164,63.906,64.572,108.063,79.227c5.14,0.954,8.945,0.283,11.419-1.996 c2.475-2.282,3.711-5.14,3.711-8.562c0-0.571-0.049-5.708-0.144-15.417c-0.098-9.709-0.144-18.179-0.144-25.406l-6.567,1.136 c-4.187,0.767-9.469,1.092-15.846,1c-6.374-0.089-12.991-0.757-19.842-1.999c-6.854-1.231-13.229-4.086-19.13-8.559 c-5.898-4.473-10.085-10.328-12.56-17.556l-2.855-6.57c-1.903-4.374-4.899-9.233-8.992-14.559 c-4.093-5.331-8.232-8.945-12.419-10.848l-1.999-1.431c-1.332-0.951-2.568-2.098-3.711-3.429c-1.142-1.331-1.997-2.663-2.568-3.997 c-0.572-1.335-0.098-2.43,1.427-3.289c1.525-0.859,4.281-1.276,8.28-1.276l5.708,0.853c3.807,0.763,8.516,3.042,14.133,6.851 c5.614,3.806,10.229,8.754,13.846,14.842c4.38,7.806,9.657,13.754,15.846,17.847c6.184,4.093,12.419,6.136,18.699,6.136 c6.28,0,11.704-0.476,16.274-1.423c4.565-0.952,8.848-2.383,12.847-4.285c1.713-12.758,6.377-22.559,13.988-29.41 c-10.848-1.14-20.601-2.857-29.264-5.14c-8.658-2.286-17.605-5.996-26.835-11.14c-9.235-5.137-16.896-11.516-22.985-19.126 c-6.09-7.614-11.088-17.61-14.987-29.979c-3.901-12.374-5.852-26.648-5.852-42.826c0-23.035,7.52-42.637,22.557-58.817 c-7.044-17.318-6.379-36.732,1.997-58.24c5.52-1.715,13.706-0.428,24.554,3.853c10.85,4.283,18.794,7.952,23.84,10.994 c5.046,3.041,9.089,5.618,12.135,7.708c17.705-4.947,35.976-7.421,54.818-7.421s37.117,2.474,54.823,7.421l10.849-6.849 c7.419-4.57,16.18-8.758,26.262-12.565c10.088-3.805,17.802-4.853,23.134-3.138c8.562,21.509,9.325,40.922,2.279,58.24 c15.036,16.18,22.559,35.787,22.559,58.817c0,16.178-1.958,30.497-5.853,42.966c-3.9,12.471-8.941,22.457-15.125,29.979 c-6.191,7.521-13.901,13.85-23.131,18.986c-9.232,5.14-18.182,8.85-26.84,11.136c-8.662,2.286-18.415,4.004-29.263,5.146 c9.894,8.562,14.842,22.077,14.842,40.539v60.237c0,3.422,1.19,6.279,3.572,8.562c2.379,2.279,6.136,2.95,11.276,1.995 c44.163-14.653,80.185-41.062,108.068-79.226c27.88-38.161,41.825-81.126,41.825-128.906 C438.536,184.851,428.728,148.168,409.132,114.573z"
                />
              </g>
            </svg>
            <span
              style={
                Object {
                  "color": "rgba(0, 0, 0, 0.87)",
                  "fontSize": 14,
                  "fontWeight": 500,
                  "letterSpacing": 0,
                  "margin": 0,
                  "muiPrepared": true,
                  "opacity": 1,
                  "paddingLeft": 8,
                  "paddingRight": 16,
                  "position": "relative",
                  "textTransform": "uppercase",
                  "userSelect": "none",
                }
              }
            >
              Project home
            </span>
          </div>
        </div>
      </a>
    </div>
    <div
      className="menuButton"
      style={
        Object {
          "WebkitTapHighlightColor": "rgba(0,0,0,0)",
          "backgroundColor": "#ffffff",
          "borderRadius": 2,
          "boxShadow": "0 1px 6px rgba(0, 0, 0, 0.12),
                 0 1px 4px rgba(0, 0, 0, 0.12)",
          "boxSizing": "border-box",
          "color": "rgba(0, 0, 0, 0.87)",
          "display": "inline-block",
          "fontFamily": "Roboto, sans-serif",
          "margin": "10px 24px",
          "minWidth": 88,
          "muiPrepared": true,
          "position": "relative",
          "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
        }
      }
    >
      <button
        disabled={false}
        href={undefined}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        onTouchTap={[Function]}
        style={
          Object {
            "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
            "backgroundColor": "#ffffff",
            "border": 10,
            "borderRadius": 2,
            "boxSizing": "border-box",
            "cursor": "pointer",
            "display": "inline-block",
            "fontFamily": "Roboto, sans-serif",
            "fontSize": "inherit",
            "fontWeight": "inherit",
            "height": 36,
            "lineHeight": "36px",
            "margin": 0,
            "muiPrepared": true,
            "outline": "none",
            "padding": 0,
            "position": "relative",
            "textAlign": "center",
            "textDecoration": "none",
            "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
            "verticalAlign": null,
            "width": "100%",
            "zIndex": 1,
          }
        }
        tabIndex={0}
        type="button"
      >
        <div
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
        >
          <div
            style={
              Object {
                "backgroundColor": false,
                "borderRadius": 2,
                "height": 36,
                "muiPrepared": true,
                "top": 0,
                "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              }
            }
          >
            <svg
              className="menuIcon"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              style={
                Object {
                  "color": "rgba(0, 0, 0, 0.87)",
                  "display": "inline-block",
                  "fill": "rgba(0, 0, 0, 0.87)",
                  "height": 24,
                  "marginLeft": 12,
                  "marginRight": 0,
                  "muiPrepared": true,
                  "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                  "userSelect": "none",
                  "verticalAlign": "middle",
                  "width": 24,
                }
              }
              viewBox="0 0 24 24"
            >
              <path
                d="M12 5V1L7 6l5 5V7c3.31 0 6 2.69 6 6s-2.69 6-6 6-6-2.69-6-6H4c0 4.42 3.58 8 8 8s8-3.58 8-8-3.58-8-8-8z"
              />
            </svg>
            <span
              style={
                Object {
                  "color": "rgba(0, 0, 0, 0.87)",
                  "fontSize": 14,
                  "fontWeight": 500,
                  "letterSpacing": 0,
                  "margin": 0,
                  "muiPrepared": true,
                  "opacity": 1,
                  "paddingLeft": 8,
                  "paddingRight": 16,
                  "position": "relative",
                  "textTransform": "uppercase",
                  "userSelect": "none",
                }
              }
            >
              Reset game
            </span>
          </div>
        </div>
      </button>
    </div>
    <div
      className={undefined}
      disabled={undefined}
      href={undefined}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      onTouchTap={[Function]}
      style={
        Object {
          "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
          "backgroundColor": "rgb(224, 224, 224)",
          "border": 10,
          "borderRadius": 16,
          "boxShadow": null,
          "boxSizing": "border-box",
          "cursor": "default",
          "display": "flex",
          "fontFamily": "Roboto, sans-serif",
          "fontSize": "inherit",
          "fontWeight": "inherit",
          "margin": 0,
          "muiPrepared": true,
          "outline": "none",
          "padding": 0,
          "position": "relative",
          "textDecoration": "none",
          "verticalAlign": null,
          "whiteSpace": "nowrap",
          "width": "fit-content",
          "zIndex": 1,
        }
      }
      tabIndex={0}
    >
      <div
        className={undefined}
        size={32}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgb(188, 188, 188)",
            "borderRadius": "50%",
            "color": "#ffffff",
            "display": "inline-flex",
            "fontSize": 16,
            "height": 32,
            "justifyContent": "center",
            "marginRight": -4,
            "muiPrepared": true,
            "userSelect": "none",
            "width": 32,
          }
        }
      >
        <svg
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "color": "#ffffff",
              "display": "inline-block",
              "fill": "#ffffff",
              "fontSize": 19.2,
              "height": 19.2,
              "margin": 6.4,
              "muiPrepared": true,
              "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              "userSelect": "none",
              "width": 19.2,
            }
          }
          viewBox="0 0 24 24"
        >
          <path
            d="M22 5.72l-4.6-3.86-1.29 1.53 4.6 3.86L22 5.72zM7.88 3.39L6.6 1.86 2 5.71l1.29 1.53 4.59-3.85zM12.5 8H11v6l4.75 2.85.75-1.23-4-2.37V8zM12 4c-4.97 0-9 4.03-9 9s4.02 9 9 9c4.97 0 9-4.03 9-9s-4.03-9-9-9zm0 16c-3.87 0-7-3.13-7-7s3.13-7 7-7 7 3.13 7 7-3.13 7-7 7z"
          />
        </svg>
      </div>
      <span
        style={
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontSize": 14,
            "fontWeight": 400,
            "lineHeight": "32px",
            "muiPrepared": true,
            "paddingLeft": 12,
            "paddingRight": 12,
            "userSelect": "none",
            "whiteSpace": "nowrap",
          }
        }
      >
        0
        s
      </span>
    </div>
    <div
      className={undefined}
      disabled={undefined}
      href={undefined}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      onTouchTap={[Function]}
      style={
        Object {
          "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
          "backgroundColor": "rgb(224, 224, 224)",
          "border": 10,
          "borderRadius": 16,
          "boxShadow": null,
          "boxSizing": "border-box",
          "cursor": "default",
          "display": "flex",
          "fontFamily": "Roboto, sans-serif",
          "fontSize": "inherit",
          "fontWeight": "inherit",
          "margin": 0,
          "muiPrepared": true,
          "outline": "none",
          "padding": 0,
          "position": "relative",
          "textDecoration": "none",
          "verticalAlign": null,
          "whiteSpace": "nowrap",
          "width": "fit-content",
          "zIndex": 1,
        }
      }
      tabIndex={0}
    >
      <div
        className={undefined}
        size={32}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgb(188, 188, 188)",
            "borderRadius": "50%",
            "color": "#ffffff",
            "display": "inline-flex",
            "fontSize": 16,
            "height": 32,
            "justifyContent": "center",
            "marginRight": -4,
            "muiPrepared": true,
            "userSelect": "none",
            "width": 32,
          }
        }
      >
        <svg
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "color": "#ffffff",
              "display": "inline-block",
              "fill": "#ffffff",
              "fontSize": 19.2,
              "height": 19.2,
              "margin": 6.4,
              "muiPrepared": true,
              "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              "userSelect": "none",
              "width": 19.2,
            }
          }
          viewBox="0 0 24 24"
        >
          <path
            d="M9.01 14H2v2h7.01v3L13 15l-3.99-4v3zm5.98-1v-3H22V8h-7.01V5L11 9l3.99 4z"
          />
        </svg>
      </div>
      <span
        style={
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontSize": 14,
            "fontWeight": 400,
            "lineHeight": "32px",
            "muiPrepared": true,
            "paddingLeft": 12,
            "paddingRight": 12,
            "userSelect": "none",
            "whiteSpace": "nowrap",
          }
        }
      >
        0
      </span>
    </div>
  </div>
</div>
`;
